gqlalchemy/vendors/database.py

```diff
@@ -0,0 +1,316 @@
+# Copyright (c) 2016-2022 Memgraph Ltd. [https://memgraph.com]
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Iterator, List, Optional
+
+from ..connection import Connection
+from ..exceptions import GQLAlchemyError
+from ..models import (
+    Constraint,
+    Index,
+    Node,
+    Relationship,
+)
+
+
+class Database(ABC):
+    def __init__(
+        self,
+        host: str,
+        port: int,
+        username: str,
+        password: str,
+        encrypted: bool,
+        client_name: str,
+    ):
+        self._host = host
+        self._port = port
+        self._username = username
+        self._password = password
+        self._encrypted = encrypted
+        self._client_name = client_name
+        self._cached_connection: Optional[Connection] = None
+
+    @property
+    def host(self):
+        return self._host
+
+    @property
+    def port(self):
+        return self._port
+
+    def execute_and_fetch(self, query: str, connection: Connection = None) -> Iterator[Dict[str, Any]]:
+        """Executes Cypher query and returns iterator of results."""
+        connection = connection or self._get_cached_connection()
+        return connection.execute_and_fetch(query)
+
+    def execute(self, query: str, connection: Connection = None) -> None:
```

Comment:
I will leave the note for `Connection` class here as well. Type of `connection` and the concrete class(derived from `Database`) should be more tightly coupled